DevEx - Add documentation for bearer-over-mTLS user flows (#3997)
AzureAD/microsoft-identity-web: docs/design/User-flow-bound-credential-mtls-devex.md

# Bearer-over-mTLS client auth for user flows (`UseBoundCredential`)

## Summary

`UseBoundCredential: true` presents the app's certificate at the TLS layer

> (`mtlsauth.microsoft.com`) to authenticate to Entra ID, instead of a
> `client_assertion` JWT in the body. The access token is still a plain **bearer**

Reviewer: Nit: "instead of a `client_assertion` JWT" contradicts step 2 of *How it works*, which says `client_assertion` is still sent (x5c auto-enabled). Step 2 is correct — mTLS changes the channel, not the assertion. Verified on the wire in #3996: bound user-flow requests hit `mtlsauth.microsoft.com` **and** carry `client_assertion`.

```suggestion
(`mtlsauth.microsoft.com`) to authenticate to Entra ID. The `client_assertion`
JWT is still sent. The access token is still a plain **bearer**
```

Author: modified

Reviewer: This got fixed in the wrong direction — step 2 lost `client_assertion` but the Summary still says "instead of". The assertion is still sent; mTLS changes the channel, not the credential material. Wire dump from a bound OBO call on 4.87.0:

```
endpoint  = https://mtlsauth.microsoft.com/{tenant}/oauth2/v2.0/token
POST keys = assertion, client_assertion, client_assertion_type, client_id,
            client_info, grant_type, requested_token_use, scope
client_assertion_type = urn:ietf:params:oauth:client-assertion-type:jwt-bearer
client_assertion      = 2122 chars
```

Also makes "x5c auto-enabled" on line 122 dangling — x5c is a header of the assertion that line no longer mentions (and is why it's 2122 chars). Suggest restoring it there.

```suggestion
(`mtlsauth.microsoft.com`) to authenticate to Entra ID. The `client_assertion`
JWT is still sent. The access token is still a plain **bearer**
```

Author: Fixed in `1e914a5`. I removed the incorrect claim that mTLS replaces the client assertion and rewrote the summary at the IdWeb contract level. I also removed the MSAL transport/assertion implementation details from **How it works**.

token; only the app→ESTS leg changes. Maps to MSAL
`CertificateOptions.SendCertificateOverMtls`.

The knob **already ships** (abstractions 12.1.0) and already works for **app
tokens** — see the merged sample
[`daemon-app-cert-bound`](../../tests/DevApps/daemon-app/daemon-app-cert-bound).
This spec **extends the same knob to user flows** (auth code, OBO, silent).

**No new public API.** MSAL [#6009](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/6009)
makes the delegated flows honor the flag; IdWeb just needs the MSAL bump +
tests + docs/samples.

## Not to be confused with `ProtocolScheme = "MTLS_POP"`

| | `UseBoundCredential: true` | `ProtocolScheme: "MTLS_POP"` ([#3832](./msi-fic-pure-mtls-pop-devex.md)) |
|---|---|---|
| Set on | the **credential** (`ClientCredentials`) | the **downstream API** options |
| mTLS applies to | app → ESTS **client auth** | the **downstream API** call |
| Access token | **bearer** | **sender-constrained** (PoP) |

> | Credential | certificate only | certificate, MI, FIC-with-MI |

Reviewer: Not cert-only. `ConfidentialClientApplicationBuilderExtension` honors the flag for `CredentialType.SignedAssertion` too (→ `WithBoundClientAssertion`, throws `MissingTokenBindingCertificate` if the provider can't bind). Only `Secret` ignores it. As written this turns away the FIC audience.

```suggestion
| Credential | certificate, or signed assertion (FIC / MI / OIDC IdP) | certificate, MI, FIC-with-MI |
```

Line 138 needs the same correction.

Independent; may both be set.

## Flow coverage

| Flow | Entry point | Bearer-over-mTLS |
|---|---|---|
| App token (daemon) | `RequestAppToken` / `CallApiForApp` | ✅ shipped |
| Web-app sign-in | `AddMicrosoftIdentityWebApp` | ✅ via #6009 |
| On-behalf-of | `EnableTokenAcquisitionToCallDownstreamApi` | ✅ via #6009 |

> | Silent / refresh-token | `AcquireTokenSilent` | ✅ via #6009 |

Reviewer: Entry-point column mixes layers — `AcquireTokenSilent` is MSAL; the idweb dev calls `CallApiForUserAsync` / `IAuthorizationHeaderProvider.CreateAuthorizationHeaderForUserAsync`. Bogdan framed this as `GetHeaderForApp`/`GetHeaderForUser`, so aligning to those names makes the coverage claim land.

Author: Fixed in `1e914a5`. The column is now **IdWeb usage**, the app and OBO rows use `IDownstreamApi` / `IAuthorizationHeaderProvider`, and silent refresh is described as behavior handled internally rather than as a developer call to `AcquireTokenSilent`.

    "TenantId": "<tenant>",
    "ClientId": "<api-client-id>",
    "Audience": "api://<api-client-id>",
    "ClientCredentials": [
      {
        "SourceType": "StoreWithThumbprint",
        "CertificateStorePath": "CurrentUser/My",
        "CertificateThumbprint": "<thumbprint>",
        "UseBoundCredential": true
      }
    ]
  },
  "DownstreamApis": {
    "GraphAPI": {
      "BaseUrl": "https://graph.microsoft.com/v1.0",
      "Scopes": [ "https://graph.microsoft.com/.default" ]
    }
  }
}
```

## Config — web app (sign-in + call Graph)

```json
{
  "AzureAd": {
    "Instance": "https://login.microsoftonline.com/",
    "TenantId": "<tenant>",
    "ClientId": "<client-id>",
    "CallbackPath": "/signin-oidc",
    "ClientCredentials": [
      {
        "SourceType": "StoreWithThumbprint",
        "CertificateStorePath": "CurrentUser/My",
        "CertificateThumbprint": "<thumbprint>",
        "UseBoundCredential": true
      }
    ]
  },
  "DownstreamApi": {
    "BaseUrl": "https://graph.microsoft.com/v1.0",
    "Scopes": [ "user.read" ]
  }
}
```

Only delta vs. a plain-bearer config: the `"UseBoundCredential": true` line, and

> a **certificate** source (not `ClientSecret` — bearer-over-mTLS is cert-only).

Reviewer: Worth stating that this fails *silently*: `case CredentialType.Secret` just calls `WithClientSecret` and drops the flag — no log, no throw. Dev sets `UseBoundCredential: true` on a secret, sees no error, assumes mTLS. Should we log a warning in idweb?

Author: Thanks. I agree the user-visible behavior should be clear, but the internal credential dispatch and logging behavior are outside this document’s scope. I’ll state that `UseBoundCredential` has no effect on client-secret credentials. Any warning or fail-fast behavior should be tracked as a separate implementation change.

`DownstreamApi(s)` is unchanged — no `ProtocolScheme`, no extra headers.

## Wiring (unchanged from any cert app)

```csharp
// Web app
builder.Services
    .AddMicrosoftIdentityWebAppAuthentication(builder.Configuration)
    .EnableTokenAcquisitionToCallDownstreamApi()
        .AddDownstreamApi("DownstreamApi", builder.Configuration.GetSection("DownstreamApi"))
        .AddInMemoryTokenCaches();

// Web API (OBO)
builder.Services
    .AddMicrosoftIdentityWebApiAuthentication(builder.Configuration)
    .EnableTokenAcquisitionToCallDownstreamApi()
        .AddDownstreamApis(builder.Configuration.GetSection("DownstreamApis"))
        .AddInMemoryTokenCaches();
```

Then `await api.CallApiForUserAsync("GraphAPI")`. No mTLS-specific C#.

## How it works

1. `UseBoundCredential: true` → IdWeb builds the CCA with
   `WithCertificate(cert, new CertificateOptions { SendCertificateOverMtls = true })`

>    (`ConfidentialClientApplicationBuilderExtension.WithClientCredentialsAsync`).

Reviewer: Worth a sentence here: `GetApplicationKey` appends `+bound` per credential, so a bound config never reuses the unbound CCA/cache entry. That's the first question anyone flipping this flag on an existing deployment will ask.

Author: Thanks. CCA cache-key construction and the `+bound` discriminator are internal IdWeb implementation details, not a supported developer contract. We should validate bound/unbound isolation in tests, but I don’t think the internal cache-key format belongs in this developer-experience document.

2. Post-#6009, auth-code / OBO / silent route to `mtlsauth.microsoft.com`,
   present the cert at TLS (x5c auto-enabled).
3. MSAL returns a **bearer** token; IdWeb's default `MsalMtlsHttpClientFactory`
   supplies the mTLS transport — no extra wiring.

> 4. `IDownstreamApi` calls downstream with `Authorization: Bearer` as usual.

Reviewer: Missing the highest-value devex piece: how a dev confirms it's actually on. One line does it —

```suggestion
4. `IDownstreamApi` calls downstream with `Authorization: Bearer` as usual.

## Verifying it works

`AuthenticationResultMetadata.TokenEndpoint` contains `mtlsauth.` when the
credential is bound, and `login.microsoftonline.com` when it isn't. That is the
assertion the tests in #3996 make.
```

Author: Thanks. But, I don’t think we should add this here. Endpoint selection and `AuthenticationResultMetadata.TokenEndpoint` are handled by MSAL internally and are not part of the IdWeb developer contract. This document should focus on the IdWeb configuration and the bearer token returned to the downstream API.

## Work items

* Bump `Microsoft.Identity.Client` to the build containing #6009.
* Delegated-flow tests (auth-code, OBO, silent).

> * Two samples: `web-app-bound-credential`, `web-api-obo-bound-credential`.

Reviewer: Stale: the MSAL bump is done (4.87.0 already referenced) and delegated-flow tests landed in #3996 — worth linking so readers see current state.

Author: Fixed in `1e914a5`. The section now records MSAL 4.87.0 and the authorization-code, OBO, and direct refresh-token tests from #3996 as complete. Actual silent-acquisition refresh coverage, samples, and public documentation remain outstanding.

* Docs: cert-credentials page + a note in `token-binding.md` contrasting with `MTLS_POP`.

## Prerequisites

* MSAL with #6009 (user-flow bearer-over-mTLS).
* Abstractions ≥ 12.1.0 (already referenced — `UseBoundCredential` shipped in [PR #255](https://github.com/AzureAD/microsoft-identity-abstractions-for-dotnet/pull/255)).
* A **certificate** credential (other credential types ignore the flag).

> * Cert registered on the app; tenant/app allow-listed for mTLS client auth (else `AADSTS700027`).

Reviewer: `mtlsauth.microsoft.com` is commercial-only. Add Gov/China endpoints (or state explicitly they're out of scope) — the `MTLS_POP` devex has that table and readers will expect parity.
